fix shm size handling

runcom opened the pull request:
Much like `MemorySwappiness` is currently working to have a default..
Added a bunch of test to make sure default is working correctly

Fix #18172

Signed-off-by: Antonio Murdaca runcom@redhat.com

runcom committed:
fix shm size handling

Signed-off-by: Antonio Murdaca <runcom@redhat.com>

thaJeztah commented on daemon/container_unix.go:
```diff
@@ -1358,11 +1358,16 @@ func (daemon *Daemon) setupIpcDirs(container *Container) error {
-		if container.hostConfig.ShmSize <= 0 {
-			container.hostConfig.ShmSize = 67108864
+		// this is to handle an omitted HostConfig passed to the API which won't hit
+		// adaptContainerSettings in daemon_unix.go and won't have a default ShmSize
```
can you add back a mention that we're setting it to `64MB` by default? prevents people from scratching their heads what the magic number `67108864` is :smile:

runcom replied:
Yup will do

runcom commented:
ping @calavera @cpuguy83 @LK4D4

estesp commented:
Slightly related to @thaJeztah's comment, I wonder if it would make sense to have a constant for `DEFAULT_SHM_SIZE` == 64MB in bytes. Pretty minor as maybe there isn't a great place that all 3 locations where you utilize it could reference it (+ testcases I guess), but that way if it ever needed to change it would be less impact across the code..

Other than that, which I'm not going to be hung up on, LGTM..

runcom commented:
@estesp I thought so much about that but couldn't figure out where to put it at then end :/ if anyone has suggestions pls...

thaJeztah commented:
Yup, I saw we actually had a constant in another location for that, but agreed, not sure what a logical place would be to have it in a single place :/

cpuguy83 commented:
How about runconfig.DefaultSHMSize?

thaJeztah commented:
perhaps a separate `runconfig.defaults` or would that be cumbersome (thinking out loud®)

runcom commented:
> How about runconfig.DefaultSHMSize?

+1 to this, I'll update the code and see how it goes

runcom commented:
@cpuguy83 @estesp @thaJeztah updated with a const in runconfig

vdemeester commented:
@runcom needs a rebase :sweat:

runcom commented:
rebased

thaJeztah commented on runconfig/parse.go:
```diff
@@ -42,6 +42,9 @@ var (
 )
 
+// DefaultSHMSize is the default size of the SHM which will be mounted in the container
+const DefaultSHMSize int64 = 67108864
```
I'd suggest to mention the `64MB` here (so that people don't have to calculate what this number is), and remove it in other places (will add some comments)

thaJeztah commented on api/server/router/local/image.go:
```diff
@@ -357,7 +357,12 @@ func (s *router) postBuild(ctx context.Context, w http.ResponseWriter, r *http.R
 	buildConfig.MemorySwap = httputils.Int64ValueOrZero(r, "memswap")
 	buildConfig.Memory = httputils.Int64ValueOrZero(r, "memory")
-	buildConfig.ShmSize = httputils.Int64ValueOrZero(r, "shmsize")
+	// default 64MB if not provided at all
```
`Set to default if not provided` (or remove, because `httputils.Int64ValueOrDefault()` may be descriptive enough on its own)

thaJeztah commented on daemon/container_unix.go:
```diff
@@ -1358,11 +1358,17 @@ func (daemon *Daemon) setupIpcDirs(container *Container) error {
-			container.hostConfig.ShmSize = 67108864
+		// this is to handle an omitted HostConfig passed to the API which won't hit
+		// adaptContainerSettings in daemon_unix.go and won't have a default ShmSize
+		// of 64MB
```
can be removed

thaJeztah commented on daemon/daemon_unix.go:
```diff
@@ -131,6 +131,11 @@ func (daemon *Daemon) adaptContainerSettings(hostConfig *runconfig.HostConfig, a
 		// By default, MemorySwap is set to twice the size of Memory.
 		hostConfig.MemorySwap = hostConfig.Memory * 2
 	}
+	// Make sure the default ShmSize if 64MB if none is provided
```
`Use default ShmSize is set if none is provided`?

runcom commented:
@thaJeztah thanks, updated

cpuguy83 commented on api/client/build.go:
```diff
@@ -211,15 +211,12 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
 	}
 
-	var shmSize int64 = 67108864 // initial SHM size is 64MB
+	shmSize := runconfig.DefaultSHMSize
```
Wondering, shouldn't this be empty on the client when not set and set the default on the daemon side?

runcom replied:
updated

runcom replied:
I also think I can move the const to daemon or container as well since it isn't used in runconfig anymore

estesp commented:
LGTM

cpuguy83 commented:
LGTM

runcom commented:
I'll submit a followup because I moved defaultSHMSize to daemon and removed the dependency on knowing the default SHM size inside the server/router :/ you guys were too fast lol

estesp commented:
ha..no, you were too slow!! :) :) :)

runcom commented:
ouch lol
